Migration and countersigning content updates, 11ty site, take 2 (#357)
holochain/docs-pages: src/pages/concepts/7_validation.md

3. Each type of app entry and link can contain any sort of binary data whose correctness is determined by validation rules written into the application.
4. Peer validators use those rules to analyze entries and spread news about bad actors, triggering an immune response in the network.

Holochain is the engine that allows peers to move data around, validate it, and take action based on validation results. **Your DNA is mainly a collection of functions for creating, accessing, and validating data.** The design of these functions is critical to the success of your app because they define the membranes of safety between a participant and the stuff she receives from others via the DHT. Her running DNA --- her cell --- prevents her from creating invalid data and defends her from the consequences of people's invalid data. Well-designed validation rules protect everyone, while buggy validation rules leave them vulnerable.

## 'Remembering' validation results

> Some entries can be computationally expensive to validate. In a currency app, for example, the validity of a transaction depends on the account balances of both transacting parties, which is the sum of all their prior transactions. The validity of each of those transactions depends on the account balance at the time, plus the validity of the account balance of the people they transacted with, and so on and so on. The data is deeply interconnected; you don't want to wait while the coffee shop's payment terminal interrogates half the town's economic history when you're just trying to buy a cup of coffee and you're already to work.

Author: ha, whoops, it was the 'late' one

> Some entries can be computationally expensive to validate. In a currency app, for example, the validity of a transaction depends on the account balances of both transacting parties, which is the sum of all their prior transactions. The validity of each of those transactions depends on the account balance at the time, plus the validity of the account balance of the people they transacted with, and so on and so on. The data is deeply interconnected; you don't want to wait while the coffee shop's payment terminal interrogates half the town's economic history when you're just trying to buy a cup of coffee and you're already to work.

Author: ```suggestion
Some entries can be computationally expensive to validate. In a currency app, for example, the validity of a transaction depends on the account balances of both transacting parties, which is the sum of all their prior transactions. The validity of each of those transactions depends on the account balance at the time, plus the validity of the account balance of the people they transacted with, and so on and so on. The data is deeply interconnected; you don't want to wait while the coffee shop's payment terminal interrogates half the town's economic history when you're just trying to buy a cup of coffee and you're already late for work.
```

The DHT offers a shortcut --- it remembers the validation results of existing entries. You can ask the validation authorities for the parties' previous transactions if they detected any problems. You can assume that they have done the same thing for the transaction prior to those and so on. As long as you trust a decent number of your peers to be playing by the rules, the validation results attached to the most recent entry 'proves' the validity of all the entries before it.

> For evidence of attempts to You can also consult the agent activity authority for any peer you're unsure of. They may or hold a copy of any past evidence of malicious activity, along with a record of all of the peer's source chain actions that show whether they've tried to change their history.

Author: what the heck

> For evidence of attempts to You can also consult the agent activity authority for any peer you're unsure of. They may or hold a copy of any past evidence of malicious activity, along with a record of all of the peer's source chain actions that show whether they've tried to change their history.

Author: ```suggestion
For evidence of attempts to fork a chain, you can also consult the agent activity authority for any peer you're unsure of. They hold a record of all of the peer's source chain actions that show whether they've tried to change their history.
```

It seems that agent activities will probably not act as clearinghouses for warrants, so I'll only make claims about fork attempts.

## The lifecycle(s) of a validation

Validation functions are called in two different scenarios, each with different consequences:

* When an agent first authors a record and attempts to produce DHT operations from it, and
* When an authority receives an operation for validation.

We'll carry on with the DHT illustrations from chapter 4 to show what happens when data is written, but let's add a simple validation rule: the `word` entry type has a validation rule that says that it can't contain spaces.

### Authoring

When you **commit a record**, your Holochain conductor is responsible for making sure you're playing by the rules. This protects you from publishing any invalid data that could make you look like a bad actor.

#### Valid entry

::: coreconcepts-storysequence
![](/assets/img/concepts/7.2-commit.png){.sz80p} {.center}

> Alice calls the `publish_word` zome function with the string `"eggplant"`. The function commits that word to her source chain. The conductor 'stages' the commit in the function's scratch space and returns the creation action's record hash to the `publish_word` function. The function continues executing and passes a return value back to the conductor, which holds onto it for now.

Author: You used to be able to define validation functions like `validate_create_word`, which would follow a sort of cascade of specificity -- that is, for the act of publishing a `word` entry, the `validate` function (which validates all actions) would be called first, then `validate_create` (which validates only `CreateEntry` actions) next, followed by `validate_create_word` (which validates only `CreateEntry` actions that operate on `word`s). That complexity has disappeared, along with the idea of validating actions -- now it's ops that are validated.

tl;dr: `publish_word` should still be the name of the zome function.

If an agent is committing a record to their source chain that depends on DHT data, it's **their job to make sure those references exist at commit time and explicitly reference them**. The validation function doesn't need to revalidate those dependencies, though; Holochain will use its own heuristics to determine the trustworthiness of other validators' claims when it retrieves them.

Soft things which normally require human discretion, like content moderation and code-of-conduct enforcement, are also challenging to encode unambiguously, especially when they lack the context in which the interaction is happening. In the future, coordinator zomes will be able to create revocable app-level warrants that can be used to non-definitively block an agent from a network. This is also useful for non-threat scenarios, for instance in cases where an employee resigns from a company on good terms and needs to be removed from the company's applications.

## Final note: genesis self-check

> There is one more pseudo-validation function you should know about: the **genesis self-check**. As with validation functions, you define this callback in your integrity zomes. Its job is to 'pre-validate' an agent's membrane proof before she joins a network, to prevent her from accidentally committing a membrane proof that woud bar her from the network forever.

Author: ```suggestion
There is one more pseudo-validation function you should know about: the **genesis self-check**. As with validation functions, you define this callback in your integrity zomes. Its job is to 'pre-validate' an agent's membrane proof before she joins a network, to prevent her from accidentally committing a membrane proof that would bar her from participating in the network forever.
```
